akka-stream/src/main/scala/akka/stream/impl/Sources.scala

```diff
@@ -136,3 +139,58 @@ private[akka] final class SourceQueueAdapter[T](delegate: SourceQueue[T]) extend
   def offer(elem: T): CompletionStage[QueueOfferResult] = delegate.offer(elem).toJava
   def watchCompletion(): CompletionStage[Done] = delegate.watchCompletion().toJava
 }
+
+/**
+ * INTERNAL API
+ */
+private[stream] final class BlockingStreamSource[T, S](create: () ⇒ S,
+                                                       readData: (S) ⇒ T,
```

Comment:
We need three functions anyway since we need to expose the actual resource, and then it looks funny to have the intuitive single-element read return an Iterator. If a single read operation can result in multiple output elements, then why should an empty iterator imply termination? That’s the same kind of conflated API as … nevermind ;-)